Fixing of correct status code for POST /v2/op/update when idPattern is used

Chandradeep-NEC opened the pull request:
Fix for Issue #3427

fgalan commented on src/lib/jsonParseV2/parseBatchUpdate.cpp:
```diff
@@ -104,7 +104,15 @@ std::string parseBatchUpdate(ConnectionInfo* ciP, BatchUpdate* burP)
+        r = oe.toJson();
+        return r;
+      }
+      else if (r != "OK")
```
If you are already detecting the attempt to use idPattern with POST /v2/op/update at the parsing stage (taking into account your modifications in src/lib/jsonParseV2/parseEntityObject.cpp) probably the modification in this file is not needed.

Chandradeep-NEC replied:
As per existing design in `src/lib/jsonParseV2/parseEntityObject.cpp` class, errors are only detected but thrown to other class and the handling of errors have been in other classes. I've also followed the same design for modification in parseEntityObject.cpp class.

There is one more point as if the error handling is removed from  `src/lib/jsonParseV2/parseBatchUpdate.cpp` then expected response is not returning i.e. it returning the below error code 400 whereas the expected code is 501.
```
400 Bad Request  
{"error":"BadRequest","description":"Unsupported idPattern"}
```

fgalan replied:
Ok. So at `src/lib/jsonParseV2/parseEntityObject.cpp` you return `Unsupported idPattern` and that is used here in `src/lib/jsonParseV2/parseBatchUpdate.cpp` to compose the final response to the client. I didn't get it in my first read ;)

Thus NTC

fgalan commented on src/lib/jsonParseV2/parseEntityObject.cpp:
```diff
@@ -52,6 +52,9 @@ std::string parseEntityObject
   std::string type  = jsonParseTypeNames[valueP->GetType()];
 
+  rapidjson::Document  document;
+  document.Parse(ciP->payload);
```
Note that the document has been already parsed. The calls sequence is as follows

parseBathUpdate() -> parseEntityVector()-> parseEntityObject()

and parseBathUpdate() is the one that does the parsing.

Thus, to avoid re-parsing here, I'd suggest to get the actionType at parseBatchUpdate() and based on it include a parameters `bool idPatternAllowed`, to be passed to parseEntityObject() in sequence, so you can use it to check if an error has to be returned or not.

Chandradeep-NEC replied:
Fixed in 4c6d654

fgalan commented on test/functionalTest/cases/3427_batch_update_with_idPattern/batch_update_DELETE_with_idPattern.test:
```diff
@@ -0,0 +1,210 @@
+# VALGRIND_READY - to mark the test ready for valgrindTestSuite.sh
+
+--NAME--
+V2 batch update REPLACE with idPattern
```
```suggestion
V2 batch update DELETE with idPattern
```

Chandradeep-NEC replied:
fixed in 4c6d654

fgalan commented on test/functionalTest/cases/3427_batch_update_with_idPattern/batch_update_DELETE_with_idPattern.test:
```diff
@@ -0,0 +1,210 @@
+# Copyright 2016 Telefonica Investigacion y Desarrollo, S.A.U
```
```suggestion
# Copyright 2020 Telefonica Investigacion y Desarrollo, S.A.U
```

:)

Chandradeep-NEC replied:
Fixed in 4c6d654

fgalan commented on test/functionalTest/cases/3427_batch_update_with_idPattern/batch_update_APPEND_with_idPattern.test:
```diff
@@ -0,0 +1,578 @@
+# VALGRIND_READY - to mark the test ready for valgrindTestSuite.sh
+
+--NAME--
+V2 batch update with idPattern
```
```suggestion
V2 batch update with idPattern (APPEND and APPEND_STRICT)
```

Chandradeep-NEC replied:
FIxed in 4c6d654

fgalan commented on test/functionalTest/cases/3427_batch_update_with_idPattern/batch_update_APPEND_with_idPattern.test:
```diff
@@ -0,0 +1,578 @@
+# Copyright 2016 Telefonica Investigacion y Desarrollo, S.A.U
```
```suggestion
# Copyright 2020 Telefonica Investigacion y Desarrollo, S.A.U
```

Chandradeep-NEC replied:
Fixed in 4c6d654

fgalan commented on test/functionalTest/cases/3427_batch_update_with_idPattern/batch_update_REPLACE_with_idPattern.test:
```diff
@@ -0,0 +1,216 @@
+# Copyright 2016 Telefonica Investigacion y Desarrollo, S.A.U
```
```suggestion
# Copyright 2020 Telefonica Investigacion y Desarrollo, S.A.U
```

Chandradeep-NEC replied:
Fixed in 4c6d654

fgalan commented on test/functionalTest/cases/3427_batch_update_with_idPattern/batch_update_UPDATE_with_idPattern.test:
```diff
@@ -0,0 +1,367 @@
+# VALGRIND_READY - to mark the test ready for valgrindTestSuite.sh
+
+--NAME--
+V2 batch update UPDATE with idPattern
```
```suggestion
V2 batch update UPDATE with idPattern (UPDATE)
```

Chandradeep-NEC replied:
Fixed in 4c6d654

fgalan commented on test/functionalTest/cases/3427_batch_update_with_idPattern/batch_update_UPDATE_with_idPattern.test:
```diff
@@ -0,0 +1,367 @@
+# Copyright 2016 Telefonica Investigacion y Desarrollo, S.A.U
```
```suggestion
# Copyright 2020 Telefonica Investigacion y Desarrollo, S.A.U
```

Chandradeep-NEC replied:
Fixed in 4c6d654

fgalan commented:
Thank you for your contribution!

I have provided some feedback in the form of comments. Some of them are using the github suggestion format, so they can be easily confirmed using the github web interface.

I especially like the good coverage you have provided in the PR, with a strong set of use cases in the new .test added. That's the way! :)

However, have a look to travis run of functional tests at https://travis-ci.org/telefonicaid/fiware-orion/jobs/643785789?utm_medium=notification&utm_source=github_status. Some previously existing .test seem to be failing. Did you run a *full run* (not only of the .test added by the PR) in your local environment before creating the PR?

fgalan commented:
In addition, note some lint errors also shown by travis:

```
./scripts/style_check_in_makefile.sh

style_check.sh: [src/lib/serviceRoutinesV2]: found no mayor errors (5 minor errors) in 5999 lines of source code in 82 files (.28% style-guide-incompatibilities)

style_check.sh: [src/lib/logSummary]: found no mayor errors (0 minor errors) in 229 lines of source code in 2 files (0% style-guide-incompatibilities)

style_check.sh: found 1 errors (1 mayor, 0 minor errors) in 6214 lines of source code in 48 files (.01% style-guide-incompatibilities)

Lint Errors:

0001 errors of category 'Missing space before ('

=================================================================

src/lib/jsonParseV2/parseEntityObject.cpp:94:  Missing space before ( in if(  [whitespace/parens] [5]

Total errors found: 1
```

fgalan commented:
Any update in the progress of this PR, please?

Thanks!

Chandradeep-NEC commented:
Work on this issue has been stopped due to other work. But now we have resumed work on this and will update you soon.

Thanks

Chandradeep-NEC commented:
> Any update in the progress of this PR, please?

I've updated the PR, please review it.

fgalan commented:
Note travis reports a fail. Looking into it, it seems is something related with linting and code style:

```
0002 errors of category 'Line ends in whitespace'

=================================================================

src/lib/jsonParseV2/parseBatchUpdate.cpp:88:  Line ends in whitespace.  Consider deleting these extra spaces.  [whitespace/end_of_line] [4]

src/lib/jsonParseV2/parseBatchUpdate.cpp:89:  Line ends in whitespace.  Consider deleting these extra spaces.  [whitespace/end_of_line] [4]
```

Please have a look and fix. Once travis goes to green again I'll have a look to the PR.

Thanks!

Chandradeep-NEC commented:
> Note travis reports a fail. Looking into it, it seems is something related with linting and code style

Fixed in 9418dcc.

>Once travis goes to green again I'll have a look to the PR.

Hi @fgalan 
I've updated the PR, Now travis is green:)

fgalan commented on src/lib/jsonParseV2/parseEntityObject.cpp:
```diff
@@ -88,6 +89,10 @@ std::string parseEntityObject
     }
     else if (name == "idPattern")
     {
+      if (idPatternAllowed == true)
+      {
+        return ERROR_NOTIMPLEMENTED;
+      }
```
I think it would be more clear in the inverse semantic for the bool variable. I men, this code to be:

```suggestion
      if (idPatternAllowed == false)
      {
        return ERROR_NOTIMPLEMENTED;
      }
```

changing in the other places to make it to work.

Chandradeep-NEC replied:
Fixed in 425b48c

fgalan commented on src/lib/jsonParseV2/parseBatchUpdate.cpp:
```diff
@@ -102,9 +110,17 @@ std::string parseBatchUpdate(ConnectionInfo* ciP, BatchUpdate* burP)
     {
       // param 4 for parseEntityVector(): attributes are allowed in payload
-      std::string r = parseEntityVector(ciP, iter, &burP->entities, true);
+      std::string r = parseEntityVector(ciP, iter, &burP->entities, idPatternAllowed, true);
```
`parseBatchUpdate()` never allow idPattern (at the present moment, I mean). Thus, I think a variable is not needed here, just to set the false value here. Or maybe I'm missing something...

That will save the modification in L88-103, that should be reverted to its original form.

Chandradeep-NEC replied:
Fixed in 425b48c

fgalan commented:
Thank for your work!

I have provided a couple coments on the PR code. In addition, please add an entry to CHANGES_NEXT_RELEASE about the fix. Suggestion:

```
- Fix: wrong 204 code was responeded (intead of 501) when using idPattern in POST /v2/op/update operation (#3427)
```

Chandradeep-NEC commented:
Hi @fgalan,
I've updated the PR.

> In addition, please add an entry to CHANGES_NEXT_RELEASE about the fix. Suggestion:

Entry regarding fix is added in CHANGES_NEXT_RELEASE.

fgalan commented on src/lib/jsonParseV2/parseBatchQuery.cpp:
```diff
@@ -91,7 +91,7 @@ std::string parseBatchQuery(ConnectionInfo* ciP, BatchQuery* bqrP)
     if (name == "entities")
     {
       // param 4 to parseEntityVector(): attributes are NOT allowed in payload
```
```suggestion
      // param 4 to parseEntityVector(): idPattern is allowed in batch queries
      // param 5 to parseEntityVector(): attributes are NOT allowed in payload
```

Chandradeep-NEC replied:
Fixed in 6daf831

fgalan commented on src/lib/jsonParseV2/parseBatchUpdate.cpp:
```diff
@@ -102,9 +102,17 @@ std::string parseBatchUpdate(ConnectionInfo* ciP, BatchUpdate* burP)
     if (name == "entities")
     {
       // param 4 for parseEntityVector(): attributes are allowed in payload
```
```suggestion
      // param 4 to parseEntityVector(): idPattern is not allowed in batch updates
      // param 5 for parseEntityVector(): attributes are allowed in payload
```

Chandradeep-NEC replied:
Fixed in 6daf831

fgalan commented:
Just only a couple of minors commit suggestions to fix some code comments and I think the PR is ready :)

Thanks for your work!

Chandradeep-NEC commented:
Hi @fgalan ,
I've updated the PR as per comments :)
